OSSM-2233: introduces xns-informer

bartoszmajsak opened the pull request:
This PR ports `xns-informer` specific changes to `upstream@1.16` version. 

On top of that following additions have been incorporated:

  - maistra does not rely on vendoring anymore
  - API types are now coming from `maistra/api` dependency rather than being generated during the build process
  - minor code cleanup (e.g. renaming `DisableNodeAccess` to `EnableNodeAccess` to be aligned with CRD scan)

**IMPORTANT**: `telemetry/stackdriver` tests has been disabled due to breaking changes in upstream (see https://github.com/maistra/istio/pull/671#issuecomment-1302545649 for details).

dgn committed:
[no-cluster-rbac] Introduce xns-informer (#533)

MAISTRA-2051: Add xns-informer for Kubernetes types (#223)

* vendor: Add github.com/maistra/xns-informer

* Update mirrored licenses to fix gen-check

* MAISTRA-2051: Add xns-informer for Kubernetes types

This integrates the xns-informer library for multi-namespace support,
but only for core Kubernetes types.  The informers are hard-coded to
watch all namespaces here.  The integration with MemberRoll will come
in a follow-up.

MAISTRA-2051 Add xns-informer for Istio types (#221)

MAISTRA-2051: Add MemberRollController (#227)

* MAISTRA-2051: Add MemberRollController

Includes the following:

* 75742b0cb4 - Add Makefile for Maistra client generation

This adds a Makefile with tasks for generating the Maistra specific
clients. It is adapted from the one in the Istio client-go repository.

* 5f8562cde4 - Add MemberRollController

Squashed commit, consisting of:

  * MAISTRA-417: Add MemberRollController

  * MAISTRA-450: Switch across to ConfiguredMembers and check existing
    config for changes

  * MAISTRA-1005 Do not expect namespaces to have a stable ordering

The clients have also been regenerated.

* 1b7a89dac6 - Add missing licenses

* fe2f8cda8f - Fix Copyright holder

* 06ea2db574 - MAISTRA-1666: Add `go mod vendor` as part of `make gen`

This is to make sure all PR's will have an updated
vendor directory. The `gencheck` prow job will ensure that.

* MAISTRA-2051: Update MemberRollController for xns-informers

* MAISTRA-2051: Add AddMemberRoll to kube client

* MAISTRA-2051: Configure MemberRoll controller in istiod

MAISTRA-2051: Add xns-informer for Service APIs (#229)

* MAISTRA-2051: Update xns-informers

* MAISTRA-2051: Add xns-informer for Service APIs

MAISTRA-2051: Use shared Kubernetes client in galley (#241)

This moves galley to the shared Kubernetes client, which will let it
use xns-informers for multi-namespace support like everything else.

MAISTRA-2051: Update xns-informers to latest (#252)

* MAISTRA-2051: Reset Kubernetes client to release-1.8 state

* MAISTRA-2051: Use xns-informers in Kubernetes client

* MAISTRA-2051: Un-skip tests with xns-informers issues

These tests were skipped because of issues with xns-informers, which
have now been solved.  They should no longer be skipped.

* MAISTRA-2051: Update xns-informers to latest

* MAISTRA-2051: Skip problematic integration test: TestDashboard

MAISTRA-2051: Use MultiNamespaceInformer in galley (#254)

* Revert "MAISTRA-2051: Use shared Kubernetes client in galley (#241)"

This reverts commit 0b1567a34d9c4e824ae694a0411997ca06345a2d.

* MAISTRA-2051: Use MultiNamespaceInformer in galley

This is a rework of the previous change that used the central
Kubernetes client in Galley.  The watchers in Galley manage stopping
and starting individual informers, which doesn't work well with the
factory returning cached informers.  This creates multi-namespace
informers individually instead of using the central factory.  It's
also a much smaller change overall and doesn't affect the tests.

* MAISTRA-2051: Add GetMemberRoll method to Kubernetes client

* MAISTRA-2051: Integrate MemberRoll with Galley

This configures the mulit-namespace aware infomers in Galley to
respond to changes in the set of namespaces via the MemberRoll.

This includes parts of the following:

  - 4b70b8dce74 MAISTRA-1895 Add option to enable/disable CRD scan
  - 6e8019eed8a MAISTRA-1968 allow analyzer to work with multi list watcher

MAISTRA-1724: Don't watch namespaces if MemberRoll is used (#257)

Rewrite of the following for Maistra 2.1 / Istio 1.8 rebase:

a161a53541 - MAISTRA-1724 Don't watch namespaces if MRC is used (#161)

MAISTRA-1755: invoke UpdateNamespaces() as part of MemberRollController.Register()

Cherry-pick of f0eed15f7a for Maistra 2.1 / Istio 1.8 rebase.

MAISTRA-1724: Better hanlde deletion in NamespaceController

MAISTRA-2153: Disable namespace informer if MemberRoll is used (#283)

This causes istiod to skip creating the namespace informer in the
Kubernetes service registry controller if a MemberRoll is specified,
because Maistra cannot read namespace objects.  This will affect the
behavior of multi-network deployments that do not use mesh networks
for configuration, i.e. a namespace cannot be labeled with a default
network now.

MAISTRA-2197: Push first update when cache is warm (#294)

Change SMMR Controller behavior to only send the first update to
listeners when it has a list of members ready. In other words, when
cache is synced.

Before this, the first update was sent with only one hardcoded member:
The control plane namespace. Now it is sent with the full list of SMMR
members.

While on that, switch to our own log scope, to facilitate debugging.

Manual cherry pick of https://github.com/maistra/istio/pull/291

MAISTRA-2233: Start xns-informers with empty namespace set (#303)

The multi-namespace informers created by xns-informers default to
watching all namespaces to match upstream behaviour.  This causes
permissions issues in Maistra because they will do this until the
first update from the MemberRoll controller sets the correct set of
namespaces on each informer.  We should instead explicitly configure
them with an empty set of namespaces at startup.

MAISTRA-2234: Seed MemberRoll listeners with system namespace (#302)

The MemberRoll controller was changed in MAISTRA-2197 to not send the
initial update with just the system namespace to all listeners.  This
can be a problem when the user hasn't created an SMMR resource yet,
because any listeners registered before the caches sync may never get
an update informing them to watch the system namespace.

This attempts to work around the issue by having all listeners
individually wait on the caches to sync, then seed just the system
namespace if no SMMR resource is found.

MAISTRA-2271: Update xns-informers to fix delete events issue (#315)

This updates xns-informers to include a fix for sending delete events
for all objects when a namespace is no longer watched.

See: https://github.com/maistra/xns-informer/pull/12

Co-authored-by: Brad Ison <brad.ison@redhat.com>

jewertow committed:
[no-cluster-rbac] MAISTRA-2051: Allow istiod to run without Node access (#255) (#545)

This is a reimplementation of the following for the Maistra 2.1 /
Istio 1.8 rebase:

 - MAISTRA-1723: Allow disabling NodePort gateway support (#180)
 - Ensure Pilot can run without privileges for reading nodes (#125)

A new istiod flag is added, `disableNodeAccess`, that will disable any
feature relying on access to Node objects.

Co-authored-by: Brad Ison <brad.ison@redhat.com>

jewertow committed:
[no-cluster-rbac] OSSM-1420: Ensure istiod can run with no cluster-scoped privileges (#513) (#548)

Co-authored-by: Marko Lukša <marko.luksa@gmail.com>

jewertow committed:
[no-cluster-rbac] OSSM-796: add SMMR integration test (#458) (#549)

jewertow committed:
[misc] Integration test fixes (#530)

* Set service subnet in kind provisioner

* Skip upstream operator test suite

* Run some StackDriver tests in a single-cluster setup only

This is necessary because the fake GCP metadata server does not
currently support multi-cluster setups.

luksa committed:
OSSM-2006 Fix multiNamespaceInformer.HasSynced()

bartoszmajsak commented:
/test integration-pilot

bartoszmajsak commented:
/test all

dgn commented on pkg/kube/client.go:
```diff
@@ -398,7 +411,12 @@ func newClientInternal(clientFactory *clientFactory, revision string) (*client,
+	c.kubeInformer = kubeinformer.NewSharedInformerFactoryWithOptions(
+		c.kube,
+		resyncInterval,
+		kubeinformer.WithNamespaces(), // Maistra needs to start with an empty namespace set.
```
This shouldnt be here anymore. I think you'll need to add another commit by Marko to this PR to align with changes in xns-informer. It's this one: https://github.com/maistra/istio/commit/26ac5592c9bbd983544db6fcab4b232ec16c480b

dgn replied:
Damn, I see that it is making changes to both ior and this :/

bartoszmajsak replied:
@dgn Is it complicating things? Can't I simply ignore `ior` bits for this part of work?

dgn commented:
Telemetry Stackdriver tests might have to be disabled, but I'm not sure-

Quic tests need investigation

Servicemesh tests are failing because of a missing import of maistra.io/api

bartoszmajsak commented:
Thanks for going over it @dgn.

> Servicemesh tests are failing because of a missing import of maistra.io/api

It is imported, but there's no `vendor` folder anymore (or yet). 

```
+ make test.integration.servicemesh.kube.presubmit
make[1]: Entering directory '/home/prow/go/src/istio.io/istio'
go test -p 1 -v -count=1 -tags=integ -vet=off ./tests/integration/servicemesh/... -timeout 30m \
 --istio.test.retries=1  --istio.test.ci --istio.test.pullpolicy=IfNotPresent --istio.test.work_dir=/logs/artifacts --istio.test.hub=localhost:5000 --istio.test.tag=istio-testing --istio.test.kube.config=/home/user/.kube/config --istio.test.ci --istio.test.pullpolicy=IfNotPresent --istio.test.work_dir=/logs/artifacts --istio.test.hub=localhost:5000 --istio.test.tag=istio-testing --istio.test.kube.config=/home/user/.kube/config ""--istio.test.select=,-postsubmit",-postsubmit" \
2>&1 | tee >(/usr/local/bin/go-junit-report > /logs/artifacts/junit.xml)
go: downloading maistra.io/api v0.0.0-20220301154558-8f6a12a9464b
```

But our test setup assumes CRDs sit there.

https://github.com/maistra/istio/blob/cccdbb9ed1089992171c81fb490320d8387f7704/tests/integration/servicemesh/maistra/maistra.go#L50

So we could either hook `go mod vendor` to the test process or try to embed those manifests in the `maistra/api` itself through e.g. `//go:embed` directive.

bartoszmajsak committed:
OSSM-2233: uses maistra.io/api instead of copying

    Before we were generating and copying Maistra API types, now we use it
    as a dependency.

    Additionally following changes are introduced:

     - vendoring has been removed, as we now use embedded manifests from maistra/api in tests
     - cleans up Makefiles
     - removes unnecessary files (generated API)

bartoszmajsak commented:
/test all

bartoszmajsak commented:
/test integration-servicemesh

bartoszmajsak commented:
/test integration-pilot

bartoszmajsak commented:
/test integration-servicemesh

bartoszmajsak commented:
/test istio-integration-security

bartoszmajsak commented:
/test integration-security
/test integration-telemetry
/test integration-helm

bartoszmajsak commented:
> Telemetry Stackdriver tests might have to be disabled, but I'm not sure-

@dgn Can you shed some light on why you are not sure? They're currently failing but is there some way to fix it? Why they were disabled earlier?

> Quic tests need investigation

How they can be run?

dgn commented:
> > Telemetry Stackdriver tests might have to be disabled, but I'm not sure-
> 
> @dgn Can you shed some light on why you are not sure? They're currently failing but is there some way to fix it? Why they were disabled earlier?

Sure! In the past we had to disable the stackdriver tests altogether, as they expected to be run on a GCP cluster (and we're not doing that). Then, upstream introduced a mock metadata server that basically made them work on every cluster by simulating a GCP environment. That however only worked in single-cluster setups, so we had to limit them to those (by doing `SingleCluster()` on the test suite IIRC.

> 
> > Quic tests need investigation
> 
> How they can be run?

Generally, you can run any suite locally by doing
```bash
./prow/integ-suite-kind.sh ${MAKE_TARGET}
```

and there are dynamic make targets based on directory structure. Just make sure to add `.kube` suffix. So in this case, something like:
```bash
./prow/integ-suite-kind.sh tests.integration.telemetry.kube
```

bartoszmajsak commented:
/test integration-security
/test integration-telemetry
/test integration-helm
/test integration-pilot

bartoszmajsak commented:
Here's the reason why our telemetry test might be failing. https://github.com/istio/istio/issues/35923#issuecomment-1302524396

bartoszmajsak committed:
tmp: disables strackdriver tests

bartoszmajsak commented:
/test all

bartoszmajsak commented:
> This shouldnt be here anymore. I think you'll need to add another commit by Marko to this PR to align with changes in xns-informer. It's this one: https://github.com/maistra/istio/commit/26ac5592c9bbd983544db6fcab4b232ec16c480b

@dgn I only added xns-informer part, leaving ior out.

bartoszmajsak commented:
/test all

bartoszmajsak commented on pkg/kube/namespace/filter.go:
```diff
@@ -252,3 +253,69 @@ func (d *discoveryNamespacesFilter) isSelected(labels labels.Set) bool {
+
+// FilterNamespace is not implemented, because Maistra does not filter namespaces;
+// this function should never be called if memberRollName is specified, so it's safe to call panic.
+func (d *maistraDiscoveryNamespacesFilter) FilterNamespace(nsMeta metav1.ObjectMeta) bool {
```
This is a new function that wasn't implemented in the previous version - this is for namespace discovery.  I think in our case it's all handled by xns-informer. Correct me if I'm wrong.

dgn replied:
correct. this is only called by the handler on the namespace informer in NamespaceController, which we're not running

jwendell commented on Makefile.core.mk:
```diff
@@ -474,3 +475,5 @@ include tools/packaging/packaging.mk
 
 include common/Makefile.common.mk
+
+include pkg/servicemesh/maistra.mk
```
At some point we moved this file to `maistra/maistra.mk`: https://github.com/maistra/istio/blob/maistra-2.3/maistra/maistra.mk

Do you think we could move it now so that next rebases don't need to do it in two steps?

bartoszmajsak replied:
Lemme have a look which commit introduced that change.

bartoszmajsak replied:
It came with https://github.com/maistra/istio/pull/606. I think we can bring it as a separate PR.

jwendell commented on pkg/servicemesh/maistra.mk:
```diff
@@ -0,0 +1,61 @@
+	@echo "done updating vendor"
+
+.PHONY: gen
+maistra-gen: maistra-gen-k8s-client vendor
```
do we need the vendor here?

bartoszmajsak replied:
No, I actually realized that today :) I will clean it up. Thanks

bartoszmajsak replied:
That's fixed now.

bartoszmajsak commented:
/test integration-servicemesh

dgn commented on pilot/pkg/config/kube/ingressv1/status.go:
```diff
@@ -61,20 +61,29 @@ func (s *StatusSyncer) Run(stopCh <-chan struct{}) {
 
 // NewStatusSyncer creates a new instance
-func NewStatusSyncer(meshHolder mesh.Holder, client kubelib.Client) *StatusSyncer {
+func NewStatusSyncer(meshHolder mesh.Holder, client kubelib.Client, noNodeAccess, enableIngressClassName bool) *StatusSyncer {
```
I know this isn't your change, but the `noNodeAccess` in combination with `enableIngressClassName` is inconsistent and bad style. I think they should be adjusted to both be `enableX`. are you okay with making that change here?

bartoszmajsak replied:
Added in 2090004883c60d4d8058969907e7942690fa4d3b

dgn commented on pilot/pkg/serviceregistry/kube/controller/controller.go:
```diff
@@ -144,6 +147,24 @@ type Options struct {
+	// watch and/or list Node objects.  If this is true, some features will not
+	// be available, e.g. NodePort gateways and determining locality information
+	// based on Nodes.
+	DisableNodeAccess bool
```
as mentioned above, I think this should be `EnableNodeAccess` (defaulting to false in our case)

jewertow replied:
I was also thinking about this while rebasing 2.3.

bartoszmajsak replied:
Yeah it's a bit counterintuitive. I will change it.

bartoszmajsak replied:
@dgn it's currently defaulted to `false` as a flag. Shouldn't it be `true` with inversed/negated name?

bartoszmajsak replied:
Additional question - does it have any implication on black-box tests where we could explicitly set this flag?

bartoszmajsak replied:
Added in 2090004883

dgn commented on pkg/servicemesh/apis/servicemesh/v1/doc.go:
```diff
@@ -0,0 +1,19 @@
+// Copyright Red Hat, Inc.
```
instead of pulling in all the files under `pkg/servicemesh/apis` and `pkg/servicemesh/client`, we might want to try to use the client code in maistra/api. Another optional improvement :wink:

bartoszmajsak replied:
Done in 4db0b35f3d

dgn reviewed (approved):
LGTM, thank you Bartosz!

I have a few comments, but all of them are related to improvements of the original commits, not so much the rebase itself. If you have time and energy to work on them I'd give you a medal :laughing: if not, I think we should file a ticket to address later and then earmark for squashing in the next rebase

bartoszmajsak commented:
Thanks a lot for the review @dgn, the suggestions are not really big so we can sneak it in here. I have two questions though:

1. Shall we create a ticket for the stackdriver tests on our end (JIRA)? They're currently disabled due to Metadata server lookup assuming tests are run on the GCP infra.
2. There are a few additional commits with my minor improvements - shall I somewhat squash them into one under OSSM-2233 so we have a bit cleaner history? Or do we simply squash all of them into one?

dgn commented:
> Thanks a lot for the review @dgn, the suggestions are not really big so we can sneak it in here. I have two questions though:
> 
>     1. Shall we create a ticket for the stackdriver tests on our end (JIRA)? They're currently disabled due to Metadata server lookup assuming tests are run on the GCP infra.

Yes, let's file a ticket. I don't know if we'll have time to work on it, but we should aim to have proper testing of StackDriver functionality. We don't want to accidentally break it.

>     2. There are a few additional commits with my minor improvements - shall I somewhat squash them into one under OSSM-2233 so we have a bit cleaner history? Or do we simply squash all of them into one?

Always +1 for squashing from me. I'm starting to think that this can all be squashed together. @jewertow @luksa thoughts?

luksa commented:
@bartoszmajsak If you're talking about squashing all the commits in this PR into a single commit, then I don't think that's a good idea. But if you just want to squash bugfix commits into the commit that introduced the buggy feature, then I'm absolutely for squashing them.

bartoszmajsak commented:
> @bartoszmajsak If you're talking about squashing all the commits in this PR into a single commit, then I don't think that's a good idea. But if you just want to squash bugfix commits into the commit that introduced the buggy feature, then I'm absolutely for squashing them.

That's exactly what I intended to do - just that I messed up last rebase so I'm recovering from `reflog` right now ;) Shouldn't take long

jwendell commented on common/scripts/kind_provisioner.sh:
```diff
@@ -178,6 +178,15 @@ function setup_kind_cluster() {
+networking:
+  # MAISTRA specific:
+  # our prow cluster uses serviceSubnet 10.96.0.0/12, so the kind cluster must use other subnet to correctly route traffic;
+  # in this case, address 10.224.0.0 is chosen randomly from available set of subnets.
```
@dgn I was wondering if it's possible/easy to change this in our prow cluster so we can get rid of this

jwendell commented on manifests/charts/istio-control/istio-discovery/templates/clusterrole.yaml:
```diff
@@ -110,6 +110,17 @@ rules:
+    resources: ["routes", "routes/custom-host"]
+    verbs: ["get", "list", "watch", "create", "delete", "update"]
+  - apiGroups: ["maistra.io"]
+    resources: ["servicemeshextensions"]
```
This resource no longer exists. It was removed in 2.3. Probably in a later commit. Do we want to keep it in this PR and remove it later or remove it now?

bartoszmajsak replied:
I'm fine with doing more in this area, however, this PR is hanging for too long and blocking subsequent changes. If removal came in later commit and tests are passing should we really do it here?

jwendell replied:
not really

jwendell commented on tests/integration/servicemesh/maistra/testdata/role.yaml:
```diff
@@ -0,0 +1,104 @@
+    resources: ["routes", "routes/custom-host"]
+    verbs: ["get", "list", "watch", "create", "delete", "update"]
+  - apiGroups: ["maistra.io"]
+    resources: ["servicemeshextensions"]
```
ditto

jwendell commented on tests/integration/telemetry/stackdriver/api/stackdriver_filter_test.go:
```diff
@@ -95,7 +95,11 @@ func TestStackdriverMonitoring(t *testing.T) {
 func TestMain(m *testing.M) {
+	// nolint: staticcheck
 	framework.NewSuite(m).
+		Skip("fake gcp metadata does not work anymore. see https://github.com/istio/istio/issues/35923#issuecomment-1302423155").
```
Do we want to track this in a Jira? If yes, please link it here

bartoszmajsak replied:
We do have an issue for it it https://issues.redhat.com/browse/OSSM-2252. Do you want to add it in the code too?

jwendell replied:
We used to, but let's not block this pr just for this.

bartoszmajsak replied:
I'll add link to the commit in the issue so at least we can easily `git revert` it.

bartoszmajsak replied:
As we advanced the target branch, I rebased my commits again so your proposed change was added :)

bartoszmajsak commented:
/retest

bartoszmajsak commented:
/retest

bartoszmajsak commented:
/retest
